src/net/dns.rs

```diff
@@ -42,33 +82,107 @@ pub(crate) async fn prune_dns_cache(context: &Context) -> Result<()> {
     Ok(())
 }
 
-/// Looks up the hostname and updates DNS cache
-/// on success.
+/// Map from hostname to IP addresses.
+///
+/// NOTE: sync mutex is used, so it must not be held across `.await`
+/// to avoid deadlocks.
+static LOOKUP_HOST_CACHE: Lazy<std::sync::Mutex<HashMap<String, Vec<IpAddr>>>> =
```

Comment:
Why not `RwLock`? Though i don't think this matter much, `mutex_guard.get(hostname).cloned()` just allocates memory and copies the vector.

Also, isn't it a problem that this cache is never cleaned?